parquet-variant/src/main/java/org/apache/parquet/variant/VariantElementConverter.java

```diff
@@ -0,0 +1,668 @@
+ * converter is responsible for finalizing the object.
+ */
+class VariantElementConverter extends GroupConverter implements VariantConverter {
+
+  // We need to remember the start position in order to tell if the child typed_value was non-null.
+  // Maybe we should just add a callback to each child's end()?
+  private int startWritePos;
+  private boolean typedValueIsObject = false;
+  private int valueIdx = -1;
+  private int typedValueIdx = -1;
+  protected VariantBuilderHolder holder;
+  protected Converter[] converters;
+
+  // The following are only used if this is an object field.
+  private String objectFieldName = null;
+  private int objectFieldId = -1;
+  private VariantObjectConverter parent = null;
+
+  // Only used if typedValueIsObject is true.
+  private Set<String> shreddedObjectKeys;
+
+  @Override
+  public void init(VariantBuilderHolder holder) {
+    this.holder = holder;
+    for (Converter converter : converters) {
+      if (converter != null) {
+        ((VariantConverter) converter).init(holder);
+      }
+    }
+  }
+
+  public VariantElementConverter(GroupType variantSchema, String objectFieldName, VariantObjectConverter parent) {
+    this(variantSchema);
+    this.objectFieldName = objectFieldName;
+    this.parent = parent;
+  }
+
+  public VariantElementConverter(GroupType variantSchema) {
+    converters = new Converter[variantSchema.getFieldCount()];
+
+    List<Type> fields = variantSchema.getFields();
+
+    for (int i = 0; i < fields.size(); i++) {
+      Type field = fields.get(i);
+      String fieldName = field.getName();
+      if (fieldName.equals("value")) {
+        this.valueIdx = i;
+        if (!field.isPrimitive() || field.asPrimitiveType().getPrimitiveTypeName() != BINARY) {
+          throw new IllegalArgumentException("Value must be a binary value");
+        }
+      } else if (fieldName.equals("typed_value")) {
+        this.typedValueIdx = i;
+      }
+    }
+
+    if (valueIdx >= 0) {
+      converters[valueIdx] = new VariantValueConverter(this);
+    }
+
+    if (typedValueIdx >= 0) {
+      Converter typedConverter = null;
+      Type field = fields.get(typedValueIdx);
+      LogicalTypeAnnotation annotation = field.getLogicalTypeAnnotation();
+      if (annotation instanceof LogicalTypeAnnotation.ListLogicalTypeAnnotation) {
+        typedConverter = new VariantArrayConverter(field.asGroupType());
+      } else if (!field.isPrimitive()) {
+        GroupType typedValue = field.asGroupType();
+        typedConverter = new VariantObjectConverter(typedValue);
+        typedValueIsObject = true;
+        shreddedObjectKeys = new HashSet<>();
+        for (Type f : typedValue.getFields()) {
+          shreddedObjectKeys.add(f.getName());
+        }
+      } else {
+        typedConverter = VariantScalarConverter.create(field.asPrimitiveType());
+      }
+
+      assert (typedConverter != null);
```

Comment:
We don't use `assert` in Java. Instead, don't initialize `typedConverter` so that the compiler detects cases where it has not been initialized. Here's an example of the error message, which happens when I change the else case so that there branches where `typedConverter` is not assigned.

![image](https://github.com/user-attachments/assets/09fe790d-56da-4d87-84b7-05316c70b698)